### PR TITLE
Added circle geometry - small feature/improvement

### PR DESCRIPTION
**Pull request by bcvery1**
Added `Circle` geometry.  Added as many parallel methods to `Circle` to match `Rect` as made sense.

All new exported functions are fully tested. Also included functions to check if a `Circle` and `Rect` intersect, e.g. collide.

### CODE REVIEW

File: geometry.go [hunk 1]
```diff
@@ -310,6 +310,183 @@ func (r Rect) Intersect(s Rect) Rect {
 	return t
 }
 
+// IntersectsCircle returns whether the Circle and the Rect intersect.
+//
+// This function will return true if:
+//  - The Rect contains the Circle, partially or fully
+//  - The Circle contains the Rect, partially of fully
+//  - An edge of the Rect is a tangent to the Circle
+func (r Rect) IntersectsCircle(c Circle) bool {
```
[the remark this answers is not in the document]
Comment (bcvery1) [reply geometry.go]: Completely agree - working on this at the moment, hopefully it won't take two long (depending how busy I am, next day or two), cheers

File: geometry.go [hunk 2]
```diff
@@ -310,6 +310,183 @@ func (r Rect) Intersect(s Rect) Rect {
...
+func (c Circle) Norm() Circle {
+	return Circle{
+		Radius: math.Abs(c.Radius),
+		Center: c.Center,
+	}
+}
+
+// Area returns the area of the Circle.
+func (c Circle) Area() float64 {
+	return math.Pi * c.Radius * 2
```
[the remark this answers is not in the document]
Comment (bcvery1) [reply geometry.go]: Oops!

File: geometry.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (faiface) [reply geometry.go]: Excellent! Take your time, I know it involves some tricky math ;)

[the remark this answers is not in the document]
Comment (bcvery1) [reply geometry.go]: @faiface I appreciate your enthusiasm for this, and more importantly your patience, I have had no time over the weekend to look at this, but I should have time this evening.  Once I'm sure it's right, you'll have it :+1:

[the remark this answers is not in the document]
Comment (faiface) [reply geometry.go]: How is this coming along?

[the remark this answers is not in the document]
Comment (bcvery1) [reply geometry.go]: I've not had the time I want to apply to this; soon.

**Comment by bcvery1** [comment]
@faiface I now believe this to be ready, please let me know if you think I've made any mistakes in the calculations!

File: geometry.go [hunk 3]
```diff
@@ -310,6 +310,239 @@ func (r Rect) Intersect(s Rect) Rect {
...
+		Center: center,
+		Radius: math.Abs(radius),
+	}
+}
+
+// IntersectsRect returns a minimal required Vector, such that moving the circle by that vector would stop the Circle
+// and the Rect intersecting.  This function returns a zero-vector if the Circle and Rect do not overlap, and if only
+// the perimeters touch.
+//
+// This function will return true if:
```
Comment (faiface) [inline geometry.go]: Won't return true, but a vector ;)

File: geometry.go [hunk 4]
```diff
@@ -310,6 +310,239 @@ func (r Rect) Intersect(s Rect) Rect {
...
+//  - The Rect contains the Circle, partially or fully
+//  - The Circle contains the Rect, partially of fully
+func (c Circle) IntersectsRect(r Rect) Vec {
```
Comment (faiface) [inline geometry.go]: Just realized that this should probably be called `IntersectRect` instead of `IntersectsRect` for consistency with the other `Intersect` method. `IntersectCircle` same.

File: geometry_test.go [hunk 5]
```diff
@@ -77,3 +78,487 @@ func TestResizeRect(t *testing.T) {
...
+			name:   "Rect.IntersectsCircle(): circle overlaps top edge",
+			fields: fields{Min: pixel.ZV, Max: pixel.V(10, 10)},
+			args:   args{c: pixel.C(pixel.V(5, 10), 1)},
+			want:   pixel.V(0, -1),
+		},
+		{
+			name:   "Rect.IntersectsCircle(): edge is tangent",
+			fields: fields{Min: pixel.ZV, Max: pixel.V(10, 10)},
+			args:   args{c: pixel.C(pixel.V(-1, 5), 1)},
+			want:   pixel.ZV,
```
Comment (faiface) [inline geometry_test.go]: Definitely need at least a few tests where both axes in the `want` vector are non-zero.

**Comment by faiface** [comment]
Alright, so I'm merging this right about now!
